sdk/core/azure-core/src/main/java/com/azure/android/core/util/CoreUtils.java

```diff
@@ -0,0 +1,177 @@
+     */
+    public static byte[] clone(byte[] source) {
+        if (source == null) {
+            return null;
+        }
+        byte[] copy = new byte[source.length];
+        System.arraycopy(source, 0, copy, 0, source.length);
+        return copy;
+    }
+
+    /**
+     * Creates a copy of the source int array.
+     * @param source Array to make copy of
+     * @return A copy of the array, or null if source was null.
+     */
+    public static int[] clone(int[] source) {
+        if (source == null) {
+            return null;
+        }
+        int[] copy = new int[source.length];
+        System.arraycopy(source, 0, copy, 0, source.length);
+        return copy;
+    }
+
+    /**
+     * Creates a copy of the source array.
+     * @param source Array being copied.
+     * @param <T> Generic representing the type of the source array.
+     * @return A copy of the array or null if source was null.
+     */
+    public static <T> T[] clone(T[] source) {
+        if (source == null) {
+            return null;
+        }
+
+        return Arrays.copyOf(source, source.length);
+    }
+
+    /**
+     * Checks if the array is null or empty.
+     * @param array Array being checked for nullness or emptiness.
+     * @return True if the array is null or empty, false otherwise.
+     */
+    public static boolean isNullOrEmpty(Object[] array) {
+        return array == null || array.length == 0;
+    }
+
+    /**
+     * Checks if the collection is null or empty.
+     * @param collection Collection being checked for nullness or emptiness.
+     * @return True if the collection is null or empty, false otherwise.
+     */
+    public static boolean isNullOrEmpty(Collection<?> collection) {
+        return collection == null || collection.isEmpty();
+    }
+
+    /**
+     * Checks if the map is null or empty.
+     * @param map Map being checked for nullness or emptiness.
+     * @return True if the map is null or empty, false otherwise.
+     */
+    public static boolean isNullOrEmpty(Map<?, ?> map) {
+        return map == null || map.isEmpty();
+    }
+
+    /**
+     * Checks if the character sequence is null or empty.
+     * @param charSequence Character sequence being checked for nullness or emptiness.
+     * @return True if the character sequence is null or empty, false otherwise.
+     */
+    public static boolean isNullOrEmpty(CharSequence charSequence) {
+        return charSequence == null || charSequence.length() == 0;
+    }
+
+    /*/**
+     * Turns an array into a string mapping each element to a string and delimits them using a coma.
+     * @param array Array being formatted to a string.
+     * @param mapper Function that maps each element to a string.
+     * @param <T> Generic representing the type of the array.
+     * @return Array with each element mapped and delimited, otherwise null if the array is empty or null.
+     * /
+    public static <T> String arrayToString(T[] array, Function<T, String> mapper) {
```

Comment:
Left this method out since it makes use of the Java 8 Stream API and it's not used by any of the interceptors in this PR. Will address this at a later stage.

Reply:
If we do not want to rewrite stuff that involves the Stream APIs, there is a couple libraries that may be worth considering, specially the last one mentioned in this [StackOverflow answer](https://stackoverflow.com/a/39515397/4967034): android-retrostreams. However, by reading the documentation I noticed it uses a GPLv2 license. I know this means we cannot look at the source code but I'm fairly certain that we could add ask authorization to add a dependency on it, I'm just not exactly sure of how this whole process works. Do you know something about this @adrianhall?

Reply:
Anything that we might expose that uses Java 8 Streams APIs wouldn't necessarily be something an Android developer would expect. I'd be fine just leaving them out. For this particular method, there are lots of other ways to accomplish the same task, so if we need it I would just write it without using Streams.

Reply:
There are a bunch of retro* projects that let you use various modern Java features on Android, but I'd shy away from using any of them. Many app developers don't want to use them, and as time goes on, more and more of them are being rendered unnecessary due to the desugaring support in Android Studio.

Reply:
1) Our core position should be that every dependency should be idiomatic to Android (i.e. not forcing Android developers into uncommon patterns), common in the ecosystem, and provide significant value. 
2) I'm not sure streams is idiomatic to android anyhow.

We probably want to look at Azure Core as "only implement what is needed to implement the core service library we need" rather than "implement a copy of azure-core from somewhere else"